src/Storages/StorageFuzzJSON.cpp

```diff
@@ -0,0 +1,511 @@
+    }
+
+    return n;
+}
+
+void fuzzJSONObject(std::shared_ptr<Node> n, WriteBuffer & out, const FuzzConfig & config, pcg64 & rnd, size_t depth)
+{
+    checkStackSize();
+    bool should_fuzz = rnd() % 100 < 100 * config.fuzzing_probability;
+
+    auto node = should_fuzz ? n : [&]() -> std::shared_ptr<Node>
+    {
+        // Create a shallow copy of the node for fuzzing.
+        auto copy = std::make_shared<Node>(*n);
+        fuzzSingleNode(*copy, config, rnd, depth);
+        return copy;
+    }();
+
+    if (node->key)
+    {
+        writeDoubleQuoted(*node->key, out);
+        out << ":";
+    }
+
+    if (node->fixed)
+    {
+        if (node->fixed->getType() == Field::Types::Which::String)
+            writeDoubleQuoted(node->fixed->get<String>(), out);
+        else
+            writeFieldText(*node->fixed, out);
+    }
+    else
+    {
+        if (!node->array && !node->nested)
+            return;
+
+        const auto & [op, cl, node_list] = node->array ? std::make_tuple('[', ']', *node->array) : std::make_tuple('{', '}', *node->nested);
+
+        out << op;
+
+        bool first = true;
+        for (const auto & ptr : node_list)
+        {
+            // TODO(jkartseva@): consider using backtraking.
+            WriteBufferFromOwnString child_out;
+            if (!first)
+                child_out << ", ";
+            first = false;
+
+            fuzzJSONObject(ptr, child_out, config, rnd, depth + 1);
+            if (out.count() + child_out.count() < config.max_output_length)
+                out << child_out.str();
+            else
+                break;
+        }
+        out << cl;
+    }
+}
+
+class FuzzJSONSource : public ISource
+{
+public:
+    FuzzJSONSource(UInt64 block_size_, Block block_header_, const FuzzConfig & config_, std::shared_ptr<Node> root_)
+        : ISource(block_header_)
+        , block_size(block_size_)
+        , block_header(std::move(block_header_))
+        , config(config_)
+        , rnd(config.random_seed)
+        , root(root_)
+    {
+    }
+    String getName() const override { return "FuzzJSON"; }
+
+protected:
+    Chunk generate() override
+    {
+        Columns columns;
+        columns.reserve(block_header.columns());
+        columns.emplace_back(createColumn());
+
+        return {std::move(columns), block_size};
+    }
+
+private:
+    ColumnPtr createColumn();
+
+    UInt64 block_size;
+    Block block_header;
+
+    StorageFuzzJSON::Config config;
+    pcg64 rnd;
+
+    std::shared_ptr<Node> root;
+};
+
+ColumnPtr FuzzJSONSource::createColumn()
+{
+    auto column = ColumnString::create();
+    ColumnString::Chars & data_to = column->getChars();
+    ColumnString::Offsets & offsets_to = column->getOffsets();
+
+    offsets_to.resize(block_size);
+    IColumn::Offset offset = 0;
+
+    for (size_t row_num = 0; row_num < block_size; ++row_num)
+    {
+        WriteBufferFromOwnString out;
+        fuzzJSONObject(root, out, config, rnd, 0);
```

Comment:
Do we want to mutate same initial json each time or it would be better to use previous result as starting point for next one to get more sophisitcated result?

Reply:
I've added a parameter that enables the reuse of a previous result.